### ISSUE TITLE
`Text.rich` does not compose its semantics label in the correct order for embedded widgets

### ISSUE THREAD

**Issue by marcianx** [flutter/flutter#111931]
b/247794358

## Steps to Reproduce

The rich text widget doesn't seem to compose the semantics labels in correct order. It seems to concatenate all the `TextSpan`s, replacing an embedded `WidgetSpan` with u+fffc (unicode object replacement character), and concatenates the `WidgetSpan` label at the end. Here's a test reproducing the issue:

```dart
testWidgets('semantics label not in order', (tester) async {
  await tester.pumpWidget(
    Directionality(
      textDirection: TextDirection.ltr,
      child: Text.rich(
        TextSpan(
          children: [
            TextSpan(text: 'before '),
            WidgetSpan(
              alignment: PlaceholderAlignment.baseline,
              baseline: TextBaseline.alphabetic,
              child: Semantics(label: 'foo'),
            ),
            TextSpan(text: ' after'),
          ],
        ),
      ),
    ),
  );
  expect(tester.getSemantics(find.byType(Text)),
      matchesSemantics(label: 'before foo after'));
});
```

Error message:
```
Expected: has semantics with label: before foo after without actions:
  <content removed>
  Actual: SemanticsNode:<SemanticsNode#1(Rect.fromLTRB(0.0, 0.0, 1591.0, 893.0), label: "before ￼
after\nfoo", textDirection: ltr)>
   Which: label was: before ￼ after
          foo
```

**Expected:** The test should pass. Specifically, the semantic label should be "before foo after".

**Actual:** It's computing the semantic string as "before ￼ after\nfoo". In between "before " and " after", there is a u+fffc unicode character (object replacement character), which is not meaningful to screen readers.

**Comment by chunhtai** [comment]
mark it as P2, as this may block the customer from passing the internal a11y test, and thus block the release in the future

**Comment by chunhtai** [comment]
update: still working on the PR

**Comment by chunhtai** [comment]
still working on it. The WIP PR was blocked by an a11y guideline test which i was working on in the previous week and fixed https://github.com/flutter/flutter/pull/112492

**Comment by chunhtai** [comment]
has a pr up https://github.com/flutter/flutter/pull/113461, but it can't be merge until https://github.com/flutter/flutter/pull/110730 is merged

**Comment by chunhtai** [comment]
updates: still blocked on https://github.com/flutter/flutter/pull/110730, the pr broke several internal tests, I am actively working on them

**Comment by marcianx** [comment]
Thanks so much for the updates!

**Comment by goderbauer** [comment]
I believe this one is blocked on the reland of https://github.com/flutter/flutter/pull/116895.

**Comment by goderbauer** [comment]
(triage): The reland landed, next: need to update https://github.com/flutter/flutter/pull/113461 again.

**Comment by goderbauer** [comment]
(triage): https://github.com/flutter/flutter/pull/113461 is back in review.

**Comment by goderbauer** [comment]
(triage): We're waiting on a roll into google3.

**Comment by chunhtai** [comment]
revert https://github.com/flutter/flutter/pull/121562
